### ISSUE TITLE
Torrent hash check even when added in queued state

### ISSUE THREAD

**Issue by MutatedHero** [arvidn/libtorrent#1869]
**Please provide the following information**

libtorrent version (or branch): 1.0.11.0

platform/architecture: Windows, 64-Bit

compiler and compiler version:

client: Deluge 1.3.14

When the client is limited to allow one download simultaneously and a torrent is added it first hashes and then goes into a queued state.
The problem with this is that if you're adding torrents with the same files from multiple trackers it starts hash checking immediately causing the additionally added torrents to stop at various checked state and then re-downloading the same data once again.

Currently: Add torrent > Hash-check on files > Queued > Download
Expected: Add torrent > Queued > Hash-check on files > Download

If there a reason for the current behavior or is it something that could be changed or made configurable by the user?

**Comment by arvidn** [comment]
The number of torrents in the checking state and the number of torrents downloading are controller by two separate limits. see http://libtorrent.org/reference-Settings.html#active_checking

That means that at least one torrent is always allowed to be checking, and in your case, one torrent is always allowed to be downloading. The most straightforward way of implementing the behavior you want would probably be to adjust those limits dynamically.

Whenever there is a downloading torrent, set the checking limit to 0, and as soon as it finishes, reset it to 1. And do the same thing for downloading torrents. Keep in mind that torrents with resume data will not necessarily go into the checking state, but will briefly be "checking resume data".

**Comment by MutatedHero** [comment]
For now I created a script to workaround the issue. In Deluge I have the ability to run a script when a torrent is added. Now the following happens:
Add torrent in paused state > Check if same name torrent exist with status "Download" or "Checking" > If true, wait 30 seconds and try again > When none of above is found, recheck torrent and resume.

If something like this was implemented I think most would see it as beneficial. That is, if it's possible to write the loop so that it's not to resource intensive. For example, if the torrent added finds a matching torrent name it then waits for the downloading torrent to send a status that it has completed the download.

**Comment by arvidn** [comment]
let me understand the issue. You add a torrent, and you expect it to be checked, but instead it is queued. This is presumably because there's another torrent already being checked that it needs to wait for.

If not, what makes it ever leave the queued state and start hash checking?

You can change this behavior to always hash check all torrents that need to in parallel, before being eligible to be started (and queued for downloading). Just increase the ``active_checking`` to some large value.

Or is that not the behavior you want?

**Comment by Seeker2** [comment]
On qBitTorrent, a hash check makes a torrent active that does tracker updates and can download the torrent while trying to do a hash check:  https://github.com/qbittorrent/qBittorrent/issues/6137
Links at the end of that to similar problems...

**Comment by MutatedHero** [comment]
@Seeker2 This does not seem to be related to the issue I'm trying to describe.

@arvidn Let's see if I can make a better description of the issue than in the first post.

**My initial problem**
I add a torrent from Tracker1 to my client. It starts downloading the files and all is good. One minute later I'm adding a torrent with the same name and files from Tracker2 and selected the same download path. The torrent from Tracker two now runs a hash check and reaches the amount that has currently been downloaded by the torrent from Tracker1.
When the check is done on the torrent from Tracker2 it starts to download from it's checked state which is, lets say 20%. I now have two torrents downloading the same files to the same folder. Eventually one will receive an error because they try to access the same block and manual intervention is required in the form of waiting until the still working torrent to finish and then recheck the one that's not.

**My first try to work around the problem that caused me to create this issue**
I then figured that if I only allow a maximum of one torrent to download at a time, it would solve my problem. That is not the case, cause as soon as I add the torrent from Tracker2, it begins to hash check. I now have Tracker1 torrent downloading and Tracker2 torrent hash checked and queued on 20% of total download. Once Tracker1 is done Tracker2 starts to download without doing another hash check, making it download the rest of the 80% for no reason since the files are already there..

**Solution**
Looking at it like this there might be two issues that both needs to be solved for it to work when the client is setup as default.
The first could be solved by adding a check on torrent add in the form of: If a torrent is added and the client already has one with the same name and download path that currently has any other status than seeding or complete, it should enter queued state.
We also need to make sure it's not hash checked before entering the queued state, which is what's happening currently. Hash check should be done on resume and not when added as queued.

I hope this clears things up. If not then I can try to capture a video of the issue. But basically you will see the issue of you add two torrents from two tracker using the same name and download path. One will finish and the other one will not. And it's not possible to work around the problem by only allowing one download at a time.

**Comment by Seeker2** [comment]
@Seeker2 This does not seem to be related to the issue I'm trying to describe.

The problem I described prevents a possible workaround to your issue.

In qBT, if you to try to add torrents in a stopped/paused state and then attempted to hash check them before starting them...qBT (and possibly Deluge) will unpause that torrent and start downloading it while hash checking, overwriting whatever may have already downloaded in those files.

**Comment by MutatedHero** [comment]
@Seeker2 That's not something I've experienced with libTorrent in Deluge. If paused and I force a recheck I've never seen it start to download/upload.

**Comment by Seeker2** [comment]
Ok, then it's a qBitTorrent-specific bug instead of a Deluge or libtorrent bug.

**Comment by arvidn** [comment]
@MutatedHero thanks, that definitely clear things up. The fundamental problem here is that two different torrents should not download to the same files.

Even though the filenames may be the same, the two torrents have different info-hashes, which means the content of those files may be different (and it's non-trivial to determine whether they are the same or not unless pieces are the same size and aligned the same). What you're trying to do is getting into mutable-torrents territory (where files downloaded from one torrent may be re-used for another).

To "blindly" add two different torrents with the same names and files to the same download directory is not reliable. That's the purpose of the error you mention ("Eventually one will receive an error because they try to access the same block"). This error originates in libtorrent, which checks to make sure no other torrent has the same file open (at least I think it may still do that, it used to at some point at least). This check is precisely meant to catch these kinds of errors.

The consequence of downloading a new torrent over an existing one is that, if the bits in the second torrent are different, the first torrent will start seeding corrupt data.

Now, to address what you actually want to do. It sounds like you basically would like to make sure that "downloading" + "checking" is never more than 1. There is no built-in controller for that in libtorrent, and I would claim that it's an unusual scenario.

The client is free to implement any kind of queueing behavior by never setting *auto-managed*, and this feels like logic that doesn't belong within the scope of libtorrent.

@Seeker2 is there a libtorrent bug report for what you describe? or is it not believed to be a libtorrent error? If it need to be fixed in libtorrent, please file a separate ticket.

**Comment by MutatedHero** [comment]
I do not think it's unusual for users to download the same files from multiple trackers where some bytes differs but I definitely see your point.

_"The consequence of downloading a new torrent over an existing one is that, if the bits in the second torrent are different, the first torrent will start seeding corrupt data."_
This is something that I've been thinking about as well. I would be nice if the client was able to download the different blocks to a separate folder or files either in the same structure or an specified path. That's just something that would save space since you would then avoid having two copies of the same thing.

What about changing so that a torrent is not hash-checking when added in queued state as the topic states? The way I see it, nothing should be done to the torrent when it's queued thus making the check begin once the torrent is resumed.

**Comment by arvidn** [comment]
>  I would be nice if the client was able to download the different blocks to a separate folder or files either in the same structure or an specified path. That's just something that would save space since you would then avoid having two copies of the same thing.

Yes, I agree. In fact, there is some basic support for this in libtorrent. It's typically referred to as mutable torrents. iirc though, in its current form, it requires cooperation from the .torrent file creator (to hint about which torrents it may share files with). Perhaps a useful feature would be for the user to either provide this list or perform an exhaustive search. However, it still requires identical piece sizes and piece-aligned files (i.e. pad files). Mutable torrents (in its current form) is primarily a feature torrent creators can take advantage of. For instance the internet archive.

There's also no good support (or protection against) torrents file sharing the same files downloading concurrently. The assumption with mutable torrents has been that they are downloaded in sequence. (patches are welcome!)

> What about changing so that a torrent is not hash-checking when added in queued state as the topic states? The way I see it, nothing should be done to the torrent when it's queued thus making the check begin once the torrent is resumed.

The queuing in libtorrent essentially have two queues, one for being checked and one for downloading. By default only one torrent is checked at a time (making other torrents needing checking to be queued up behind it). So the flow looks something closer to this:

```
add-torrent
  |
  v
check-fastresume
  |
  v
queued-for-checking
  |
  v
checking
  |
  v
queued-for-downloading
  |
  v
downloading -+
  |          |
  v          v
seeding <-> queued-for-seeding
```

The various limits you can configure restrict the number of torrents in the "checking", "downloading" and "seeding", and torrents are queued in order to enforce that.

**Comment by Seeker2** [comment]
@Seeker2 is there a libtorrent bug report for what you describe? or is it not believed to be a libtorrent error? If it need to be fixed in libtorrent, please file a separate ticket.

Since this doesn't seem to be happening on Deluge... I don't believe my example is a libtorrent bug, but not sure...the issues are described at:
https://github.com/qbittorrent/qBittorrent/issues/6137  Tracker updates done while torrent paused...if Force Rechecking torrent data (my example)
https://github.com/qbittorrent/qBittorrent/issues/5759  (possibly related issue?) Torrents paused while still re-checking

https://github.com/qbittorrent/qBittorrent/issues/6132  2 Force recheck issues, and one of them is caused by libtorrent 1.1.x:
https://github.com/qbittorrent/qBittorrent/issues/6124  Force recheck does nothing if files are updated after adding torrent (libtorrent 1.1.x to blame?)
https://github.com/qbittorrent/qBittorrent/issues/165  Force recheck does nothing when .!qB is enabled
